src/core_plugins/kibana/public/dashboard/dashboard.js

```diff
@@ -180,104 +142,81 @@ app.directive('dashboardApp', function (Notifier, courier, AppState, timefilter,
+          $scope.kbnTopNav.close('save');
+          if (id) {
+            notify.info(`Saved Dashboard as "${dash.title}"`);
+            if (dash.id !== $routeParams.id) {
+              kbnUrl.change(
+                `${DashboardConstants.EXISTING_DASHBOARD_URL}`,
+                { id: dash.id });
+            } else {
+              docTitle.change(dash.lastSavedTitle);
+            }
+          }
+        }).catch(notify.fatal);
       };
 
+      $scope.$watchCollection('options', function (newVal, oldVal) {
+        if (!angular.equals(newVal, oldVal)) {
+          dashboardState.saveState();
+        }
+      });
+      $scope.$watch('options.darkTheme', setDarkTheme);
+
+      $scope.$watch('model.query', function () {
+        dashboardState.setQuery($scope.model.query);
+      });
+
+      $scope.$listen(timefilter, 'fetch', $scope.refresh);
+
       // update root source when filters update
       $scope.$listen(queryFilter, 'update', function () {
-        updateQueryOnRootSource();
-        $state.save();
+        dashboardState.updateFilters(queryFilter);
       });
 
       // update data when filters fire fetch event
       $scope.$listen(queryFilter, 'fetch', $scope.refresh);
 
-      $scope.getDashTitle = function () {
-        return dash.lastSavedTitle || `${dash.title} (unsaved)`;
-      };
-
-      $scope.newDashboard = function () {
-        kbnUrl.change('/dashboard', {});
-      };
+      $scope.$on('$destroy', () => {
+        dashboardState.destroy();
 
-      $scope.filterResults = function () {
-        updateQueryOnRootSource();
-        $state.save();
-        $scope.refresh();
-      };
+        // Remove dark theme to keep it from affecting the appearance of other apps.
+        setDarkTheme(false);
+      });
 
-      $scope.save = function () {
-        $state.save();
-
-        const timeRestoreObj = _.pick(timefilter.refreshInterval, ['display', 'pause', 'section', 'value']);
-
-        dash.panelsJSON = angular.toJson($state.panels);
-        dash.uiStateJSON = angular.toJson($uiState.getChanges());
-        dash.timeFrom = dash.timeRestore ? timefilter.time.from : undefined;
-        dash.timeTo = dash.timeRestore ? timefilter.time.to : undefined;
-        dash.refreshInterval = dash.timeRestore ? timeRestoreObj : undefined;
-        dash.optionsJSON = angular.toJson($state.options);
-
-        dash.save()
-          .then(function (id) {
-            stateMonitor.setInitialState($state.toJSON());
-            $scope.kbnTopNav.close('save');
-            if (id) {
-              notify.info('Saved Dashboard as "' + dash.title + '"');
-              if (dash.id !== $routeParams.id) {
-                kbnUrl.change('/dashboard/{{id}}', { id: dash.id });
-              } else {
-                docTitle.change(dash.lastSavedTitle);
-              }
-            }
-          })
-          .catch(notify.fatal);
-      };
+      function setDarkTheme(enabled) {
```

Comment:
I know this code is older than your changes, but maybe it's worth cleaning up too: According to clean code best practices, functions that take a boolean should instead be replaced with two functions that directly state their effect, as in `setDarkTheme()` and `setLightTheme()`. That would avoid confustion for the reader about what `setDarkTheme(false)` actually does. Does it apply no theme? Or the green theme? 😉